component/block_cache/block_cache.go

```diff
@@ -1098,17 +1101,22 @@ func (bc *BlockCache) getOrCreateBlock(handle *handlemap.Handle, offset uint64)
 			block.flags.Clear(BlockFlagSynced)
 		} else if block.flags.IsSet(BlockFlagDownloading) {
 			log.Debug("BlockCache::getOrCreateBlock : Waiting for download to finish for committed block %v for %v=>%s", block.id, handle.ID, handle.Path)
-			<-block.state
+			_, ok := <-block.state
+			if ok {
+				block.Unblock()
+			}
 			block.flags.Clear(BlockFlagDownloading)
```

Comment:
This line we can move up, generally we are doing unblock after all flags are reset.